src/lsp-server.ts

```diff
@@ -243,6 +257,26 @@ export class LspServer {
         return rootPath;
     }
 
+    private findClosestTag(tags: any[], fileName: string, line: number) {
+        // Priority for shoosing tags (high to low):
+        // 1. if line number equal or less than given line
+        // 2. absolute distance between two lines
+        // 3. if in the same file
+        const inFileTags = tags.filter(tag => tag.file = fileName);
+        if (inFileTags.length !== 0) {
+            inFileTags.sort(function(l, r) {
```

Comment:
Since the sort results are only appliable for current `l.lineNumber`, we can't cache the sort results. Sorting all the tags for every hover is time-consuming. How about maintaining two min value, `Min(l.lineNumber - tag.line), tag.line <= l.lineNumber` and `Min(tag.line - l.lineNumber), for tag.line > l.lineNumber`, so we can compute the best candidate for just only one iteration.
```shell

                      tag
                      tag
                      tag <----- Min(l.lineNumber - tag.line)
------------hover occurred line---------- 
                      tag <----- Min(tag.line - l.lineNumber)
                      tag
                      tag
```